lib/segment/src/vector_storage/query/feedback_query.rs

```diff
@@ -0,0 +1,153 @@
+        [&self.positive, &self.negative].into_iter()
+    }
+}
+
+/// Trained coefficients for the formula. Specific to a triplet of dataset-smallmodel-bigmodel.
+#[derive(Debug, Clone, PartialEq, Serialize, Hash)]
+pub struct TrainedCoefficients {
+    pub a: OrderedFloat<f32>,
+    pub b: OrderedFloat<f32>,
+    pub c: OrderedFloat<f32>,
+}
+
+/// Query for relevance feedback scoring
+#[derive(Debug, Clone, PartialEq, Serialize, Hash)]
+pub struct FeedbackQuery<T> {
+    /// The original query vector.
+    pub query: T,
+
+    /// Pairs of results with higher difference in their golden score.
+    pub feedback_pairs: Vec<FeedbackPair<T>>,
+
+    /// Trained coefficients for the formula.
+    pub coefficients: TrainedCoefficients,
+}
+
+impl<T> FeedbackQuery<T> {
+    pub fn new(
+        query: T,
+        feedback_pairs: Vec<FeedbackPair<T>>,
+        coefficients: TrainedCoefficients,
+    ) -> Self {
+        Self {
+            query,
+            feedback_pairs,
+            coefficients,
+        }
+    }
+
+    pub fn flat_iter(&self) -> impl Iterator<Item = &T> {
+        self.feedback_pairs
+            .iter()
+            .flat_map(|pair| pair.iter())
+            .chain(std::iter::once(&self.query))
+    }
+}
+
+impl<T, U> TransformInto<FeedbackQuery<U>, T, U> for FeedbackQuery<T> {
+    fn transform<F>(self, mut f: F) -> OperationResult<FeedbackQuery<U>>
+    where
+        F: FnMut(T) -> OperationResult<U>,
+    {
+        let Self {
+            query,
+            feedback_pairs,
+            coefficients,
+        } = self;
+        Ok(FeedbackQuery::new(
+            f(query)?,
+            feedback_pairs
+                .into_iter()
+                .map(|pair| pair.transform(&mut f))
+                .try_collect()?,
+            coefficients,
+        ))
+    }
+}
+
+#[inline]
+fn linear_formula(
+    score: f32,
+    confidence: f32,
+    delta: f32,
+    coefficients: &TrainedCoefficients,
+) -> f32 {
+    let TrainedCoefficients {
+        a: OrderedFloat(a),
+        b: OrderedFloat(b),
+        c: OrderedFloat(c),
+    } = coefficients;
+
+    a * score + confidence.powf(*b) * c * delta
+}
+
+impl<T> Query<T> for FeedbackQuery<T> {
+    fn score_by(&self, similarity: impl Fn(&T) -> ScoreType) -> ScoreType {
+        let Self {
+            query,
+            feedback_pairs,
+            coefficients,
+        } = self;
+
+        let mut rescore = 0.0;
+        for (idx, pair) in feedback_pairs.iter().enumerate() {
+            let FeedbackPair {
+                positive,
+                negative,
+                confidence,
+            } = pair;
+
+            // Only use the score if it is the first feedback pair
+            let score = if idx == 0 { similarity(query) } else { 0.0 };
+
+            let delta = similarity(positive) - similarity(negative);
+
+            rescore += linear_formula(score, confidence.0, delta, coefficients);
+        }
+
+        rescore
+    }
+}
```

Comment:
This is the important section of this PR

Reply:
ok I have made it easier to implement other formulas in https://github.com/qdrant/qdrant/pull/7359/commits/ab349b5e9aa441f767c81b73cfc913c2c264fb24